projects/batfish-common-protocol/src/test/java/org/batfish/storage/FileStorageImplTest.java

```diff
@@ -0,0 +1,287 @@
+    _thrown.expect(BatfishException.class);
+    _thrown.expectMessage(
+        String.format(
+            "Environment '%s' already exists for container '%s' testrig '%s'",
+            environment.getName(), "testcontainer", "testtestrig"));
+
+    //Should throw exception for duplicate environment
+    _storage.saveEnvironment("testcontainer", "testtestrig", environment);
+  }
+
+  @Test
+  public void testUpdateEnvironment() {
+    Environment environment = getTestEnvironment("env");
+    Environment savedEnvironment =
+        _storage.saveEnvironment("testcontainer", "testtestrig", environment);
+    savedEnvironment.getEdgeBlacklist().add(new Edge("node3", "int3", "node4", "int4"));
+    Environment updatedEnvironment =
+        _storage.updateEnvironment("testcontainer", "testtestrig", savedEnvironment);
+
+    Assert.assertThat(savedEnvironment, equalTo(updatedEnvironment));
+  }
+
+  @Test
+  public void testInvalidUpdateEnvironment() {
+    Environment environment = getTestEnvironment("env");
+    _thrown.expect(BatfishException.class);
+    _thrown.expectMessage(
+        String.format(
+            "Environment '%s' doesn't exist for container '%s'->testrig '%s'",
+            environment.getName(), "testcontainer", "testtestrig"));
+
+    //test that update throws for non-existent environment
+    _storage.updateEnvironment("testcontainer", "testtestrig", environment);
+  }
+
+  @Test
+  public void testDeleteEmptyEnvironment() {
+    String containerName = "testcontainer";
+    String testrigName = "testtestrig";
+    String envName = "testenv";
+    Environment environment = Environment.builder().setName(envName).build();
+    _storage.saveEnvironment(containerName, testrigName, environment);
+
+    assertTrue(_storage.deleteEnvironment(containerName, testrigName, envName, false));
+
+    _thrown.expect(BatfishException.class);
+    _thrown.expectMessage(
+        String.format(
+            String.format(
+                "Environment '%s' doesn't exist for container '%s'->testrig '%s'",
+                envName, containerName, testrigName)));
+
+    //throws since the environment is not found in the container
+    _storage.getEnvironment(containerName, testrigName, envName);
+  }
+
+  @Test
+  public void testForceDeleteEnvironment() {
+    String containerName = "testcontainer";
+    String testrigName = "testtestrig";
+    String envName = "testenv";
+    Environment environment = getTestEnvironment(envName);
+    _storage.saveEnvironment(containerName, testrigName, environment);
+    _thrown.expect(BatfishException.class);
+    _thrown.expectMessage(String.format("'%s' is not empty, deletion must be forced", envName));
+
+    //Should throw for non empty environment
+    _storage.deleteEnvironment(containerName, testrigName, envName, false);
+
+    //Deletion should be possible with force=true
+    assertTrue(_storage.deleteEnvironment(containerName, testrigName, envName, true));
```

Comment:
This line is actually not reachable, will not run codes in same test after throws exception